adapter/service/parameters/service.go

```diff
@@ -0,0 +1,171 @@
+	cfg *Config,
+) Service {
+	cfg.Specification.Name = Parameters
+	cfg.Specification.EnsureInterfaces(requiredInterfaces()...)
+
+	return &service{
+		Service:           adapter.NewService(publisher, cfg.Specification),
+		controller:        cfg.Controller,
+		lock:              &sync.Mutex{},
+		reportingCache:    cache.NewReportingCache(),
+		reportingStrategy: cache.ReportOnChangeOnly(),
+	}
+}
+
+// service is a private implementation of a parameters FIMP service.
+type service struct {
+	adapter.Service
+
+	lock              *sync.Mutex
+	controller        Controller
+	reportingCache    cache.ReportingCache
+	reportingStrategy cache.ReportingStrategy
+}
+
+func (s *service) SetParameter(p Parameter) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if err := p.Validate(); err != nil {
+		return fmt.Errorf("%s: failed to validate parameter: %w", s.Name(), err)
+	}
+
+	spec, err := s.getParameterSpecification(p.ID)
+	if err != nil {
+		return fmt.Errorf("%s: failed to get parameter specification: %w", s.Name(), err)
+	}
+
+	if spec.ReadOnly {
+		return fmt.Errorf("%s: parameter is read only", s.Name())
+	}
+
+	if err = spec.ValidateParameter(p); err != nil {
+		return fmt.Errorf("%s: failed to validate parameter: %w", s.Name(), err)
+	}
+
+	if err = s.controller.SetParameter(p); err != nil {
+		return fmt.Errorf("%s: failed to set parameter ID %s: %w", s.Name(), p.ID, err)
+	}
+
+	return nil
+}
+
+func (s *service) SendParameterReport(id string, force bool) (bool, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	parameter, err := s.controller.GetParameter(id)
+	if err != nil {
+		return false, fmt.Errorf("%s: failed to get parameter ID %s: %w", s.Name(), id, err)
+	}
+
+	if !force && !s.reportingCache.ReportRequired(s.reportingStrategy, EvtParamReport, "", parameter) {
+		return false, nil
+	}
+
+	message := fimpgo.NewObjectMessage(
+		EvtParamReport,
+		s.Name(),
+		parameter,
+		nil,
+		nil,
+		nil,
+	)
+	message.WithStorageStrategy(fimpgo.StorageStrategyAggregate, parameter.ID)
+
+	if err = s.SendMessage(message); err != nil {
+		return false, fmt.Errorf("%s: failed to send parameter report: %w", s.Name(), err)
+	}
+
+	return true, nil
+}
+
+func (s *service) SendSupportedParamsReport(force bool) (bool, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	parameters, err := s.controller.GetParameterSpecifications()
+	if err != nil {
+		return false, fmt.Errorf("%s: failed to get supported parameters: %w", s.Name(), err)
+	}
+
+	if !force && !s.reportingCache.ReportRequired(s.reportingStrategy, EvtSupParamsReport, "", parameters) {
+		return false, nil
+	}
+
+	message := fimpgo.NewObjectMessage(
+		EvtSupParamsReport,
+		s.Name(),
+		parameters,
+		nil,
+		nil,
+		nil,
+	)
+
+	if err = s.SendMessage(message); err != nil {
+		return false, fmt.Errorf("%s: failed to send supported parameters report: %w", s.Name(), err)
+	}
+
+	return true, nil
+}
+
+func (s *service) getParameterSpecification(id string) (ParameterSpecification, error) {
```

Comment:
Why return param is a copy instead of a pointer?

Reply:
Fixed